### PR TITLE
Refactor/download function

### PR DESCRIPTION
**Pull request by dbampalikis**
This is a PR in preparation for the filisize module, which will return the size of the files. Some of the code is the same among the two modules, therefore, some functions were extracted to avoid duplication.

### CODE REVIEW

**Review by shreyasshivakumara — APPROVED** [review]
Looks good! :)

**Review by aaperis — APPROVED** [review]
looks good :-)

File: download/download_test.go [hunk 1]
```diff
@@ -109,3 +109,39 @@ func (suite *TestSuite) TestCreateFilePath() {
...
+	assert.NoError(suite.T(), err)
+
+	// Folder URL does not exist
+	fileLocation := "https://some/base/A352744B-2CB4-4738-B6B5-BA55D25FB469/some/"
+
+	urlsFilePath, err := GetURLsListFile(currentPath, fileLocation)
+	assert.Equal(suite.T(), urlsFilePath, "")
+	// The error differs locally and in the repo, therefore checking that error exists
+	// In github the error produced is
+	//assert.EqualError(suite.T(), err, "failed to download file, reason: Get \"https://some/base/A352744B-2CB4-4738-B6B5-BA55D25FB469/some/urls_list.txt\": dial tcp: lookup some: Temporary failure in name resolution")
```
Comment (kostas-kou) [inline download/download_test.go]:
```suggestion
	// assert.EqualError(suite.T(), err, "failed to download file, reason: Get \"https://some/base/A352744B-2CB4-4738-B6B5-BA55D25FB469/some/urls_list.txt\": dial tcp: lookup some: Temporary failure in name resolution")
```

File: download/download_test.go [hunk 2]
```diff
@@ -109,3 +109,39 @@ func (suite *TestSuite) TestCreateFilePath() {
...
+	// Locally the error varies with the following being one alternative
+	//assert.EqualError(suite.T(), err, "failed to download file, reason: Get \"https://some/base/A352744B-2CB4-4738-B6B5-BA55D25FB469/some/urls_list.txt\": dial tcp: lookup some: no such host")
```
Comment (kostas-kou) [inline download/download_test.go]:
```suggestion
	// assert.EqualError(suite.T(), err, "failed to download file, reason: Get \"https://some/base/A352744B-2CB4-4738-B6B5-BA55D25FB469/some/urls_list.txt\": dial tcp: lookup some: no such host")
```

**Review by kostas-kou — APPROVED** [review]
Looks good. I just added a couple of minor changes because linter was complaining

File: download/download.go [hunk 3]
```diff
@@ -104,14 +104,44 @@ func getURLsFile(urlsFilePath string) (urlsList []string, err error) {
...
 	}
-
 	if len(urlsList) == 0 {
 		return urlsList, fmt.Errorf("failed to get list of files, empty file")
 	}
 
 	return urlsList, scanner.Err()
 }
 
+// GetURLsListFile is returning the path to the urls_list.txt by haldning the URL
```
Comment (jbygdell) [inline download/download.go:114]:
```suggestion
// GetURLsListFile is returning the path to the urls_list.txt by handling the URL
```

File: download/download_test.go [hunk 4]
```diff
@@ -109,3 +109,39 @@ func (suite *TestSuite) TestCreateFilePath() {
...
+	// File URL does not exist
+	fileLocation = "https://some/base/A352744B-2CB4-4738-B6B5-BA55D25FB469/some/urls_list.txt"
+
+	urlsFilePath, err = GetURLsListFile(currentPath, fileLocation)
+	assert.Equal(suite.T(), urlsFilePath, "")
+	// The error differs locally and in the repo, therefore checking that error exists
+	// In github the error produced is
+	// assert.EqualError(suite.T(), err, "failed to download file, reason: Get \"https://some/base/A352744B-2CB4-4738-B6B5-BA55D25FB469/some/urls_list.txt\": dial tcp: lookup some: Temporary failure in name resolution")
+	// Locally the error varies with the following being one alternative
+	// assert.EqualError(suite.T(), err, "failed to download file, reason: Get \"https://some/base/A352744B-2CB4-4738-B6B5-BA55D25FB469/some/urls_list.txt\": dial tcp: lookup some: no such host")
```
Comment (jbygdell) [inline download/download_test.go:139]: you could just assert that the call returns an error that starts with "failed to download file, reason:"

File: download/download_test.go [hunk 5]
```diff
@@ -109,3 +109,39 @@ func (suite *TestSuite) TestCreateFilePath() {
...
+	fileLocation := "https://some/base/A352744B-2CB4-4738-B6B5-BA55D25FB469/some/"
+
+	urlsFilePath, err := GetURLsListFile(currentPath, fileLocation)
+	assert.Equal(suite.T(), urlsFilePath, "")
+	// The error differs locally and in the repo, therefore checking that error exists
+	// In github the error produced is
+	// assert.EqualError(suite.T(), err, "failed to download file, reason: Get \"https://some/base/A352744B-2CB4-4738-B6B5-BA55D25FB469/some/urls_list.txt\": dial tcp: lookup some: Temporary failure in name resolution")
+	// Locally the error varies with the following being one alternative
+	// assert.EqualError(suite.T(), err, "failed to download file, reason: Get \"https://some/base/A352744B-2CB4-4738-B6B5-BA55D25FB469/some/urls_list.txt\": dial tcp: lookup some: no such host")
+	assert.NotEqual(suite.T(), err, nil)
```
Comment (jbygdell) [inline download/download_test.go:128]: If you want to be fancy and have a test that will pass you can use the `net/http/httptest` package

**Review by jbygdell — CHANGES_REQUESTED** [review]
Variable names are a bit unintuitive.

Test should also contain a part that is not some kind of failure.

File: download/download_test.go [hunk 6]
```diff
@@ -109,3 +127,85 @@ func (suite *TestSuite) TestCreateFilePath() {
...
+
+	// Check that the file contains the correct urls
+	expectedUrls, err = ioutil.ReadFile(urlsFilePath)
+	assert.NoError(suite.T(), err)
+	assert.Equal(suite.T(), expectedUrls, []byte(urlsList))
+
+	// Remove the file created from the downloadFile function
+	_ = os.Remove(urlsFilePath)
+
+}
```
Comment (jbygdell) [inline download/download_test.go:211]: This could have been one test, but eh, the most important is that we test all scenarios.

**Review by jbygdell — APPROVED** [review]

**Pull request merged**